core/src/main/python/synapse/ml/llm/HuggingFaceCausallmTransform.py

```diff
@@ -0,0 +1,373 @@
+        param = self.getModelParam().get_param()
+        inputs = tokenizer(prompt, return_tensors="pt").input_ids
+        outputs = model.generate(inputs, **param)
+        decoded_output = tokenizer.batch_decode(outputs, skip_special_tokens=True)[0]
+        return decoded_output
+
+    def _predict_single_chat(self, prompt, model, tokenizer):
+        param = self.getModelParam().get_param()
+        chat = [{"role": "user", "content": prompt}]
+        formatted_chat = tokenizer.apply_chat_template(
+            chat, tokenize=False, add_generation_prompt=True
+        )
+        tokenized_chat = tokenizer(
+            formatted_chat, return_tensors="pt", add_special_tokens=False
+        )
+        inputs = {
+            key: tensor.to(model.device) for key, tensor in tokenized_chat.items()
+        }
+        merged_inputs = {**inputs, **param}
+        outputs = model.generate(**merged_inputs)
+        decoded_output = tokenizer.decode(
+            outputs[0][inputs["input_ids"].size(1) :], skip_special_tokens=True
+        )
+        return decoded_output
+
+    def _transform(self, dataset):
+        """Transform method to apply the chat model."""
+
+        def _process_partition(iterator, task):
+            """Process each partition of the data."""
+            peekable_iterator = Peekable(iterator)
+            try:
+                first_row = peekable_iterator.peek()
+            except StopIteration:
+                return None
+
+            model, tokenizer = self.load_model()
+
+            for row in peekable_iterator:
+                prompt = row[self.getInputCol()]
+                if task == "chat":
+                    result = self._predict_single_chat(prompt, model, tokenizer)
+                elif task == "complete":
+                    result = self._predict_single_complete(prompt, model, tokenizer)
+                row_dict = row.asDict()
+                row_dict[self.getOutputCol()] = result
+                yield Row(**row_dict)
+
+        input_schema = dataset.schema
+        output_schema = StructType(
+            input_schema.fields + [StructField(self.getOutputCol(), StringType(), True)]
+        )
+        result_rdd = dataset.rdd.mapPartitions(
+            lambda partition: _process_partition(partition, "chat")
+        )
+        result_df = result_rdd.toDF(output_schema)
+        return result_df
+
+    def complete(self, dataset):
+        input_schema = dataset.schema
+        output_schema = StructType(
+            input_schema.fields + [StructField(self.getOutputCol(), StringType(), True)]
+        )
+        result_rdd = dataset.rdd.mapPartitions(
+            lambda partition: self._process_partition(partition, "complete")
+        )
+        result_df = result_rdd.toDF(output_schema)
+        return result_df
+
+
+class Downloader:
+    def __init__(
+        self,
+        base_cache_dir="./cache",
+        base_url="https://mmlspark.blob.core.windows.net/huggingface/",
```

Comment:
lets use 

```
%sh
azcopy cp https://mmlspark.blob.core.windows.net/huggingface/blah /lakehouse/blah
```

Reply:
Youy can also put in the mardown cell a little explanation of this and how its just for a speedup otherwise it will download from the huggingface hub